mlflow/server/js/src/experiment-tracking/components/artifact-view-components/ShowArtifactTableView.js

```diff
@@ -0,0 +1,136 @@
+import { getSrc } from './ShowArtifactPage';
+import { getArtifactContent } from '../../../common/utils/ArtifactUtils';
+import { Table } from '@databricks/design-system';
+import Papa from 'papaparse';
+
+class ShowArtifactTableView extends Component {
+  constructor(props) {
+    super(props);
+    this.fetchArtifacts = this.fetchArtifacts.bind(this);
+  }
+
+  static MAX_ROW_LENGTH = 500;
+
+  static propTypes = {
+    runUuid: PropTypes.string.isRequired,
+    path: PropTypes.string.isRequired,
+    getArtifact: PropTypes.func,
+  };
+
+  static defaultProps = {
+    getArtifact: getArtifactContent,
+  };
+
+  state = {
+    loading: true,
+    error: undefined,
+    data: undefined,
+    headers: undefined,
+    originalRowLength: 0,
+    text: undefined,
+  };
+
+  componentDidMount() {
+    this.fetchArtifacts();
+  }
+
+  componentDidUpdate(prevProps) {
+    if (this.props.path !== prevProps.path || this.props.runUuid !== prevProps.runUuid) {
+      this.setState({ data: undefined, headers: undefined, originalRowLength: 0, loading: true });
+      this.fetchArtifacts();
+    }
+  }
+
+  render() {
+    if (this.state.loading) {
+      return <div className='artifact-text-view-loading'>Loading...</div>;
+    }
+    if (this.state.error) {
+      return (
+        <div className='artifact-text-view-error'>
+          Oops we couldn't load your file because of an error.
+        </div>
+      );
+    }
+
+    if (this.state.data) {
+      // eslint-disable-next-line max-len
+      const rowPreviewMessage = `Previewing the first ${this.state.data.length} rows out of ${this.state.originalRowLength}`;
+
+      const columns = this.state.headers.map((f) => ({
+        title: f,
+        dataIndex: f,
+        key: f,
+        sorter: (a, b) => {
+          if (isNaN(a)) {
+            return a[f].localeCompare(b[f]);
+          } else {
+            return a[f] - b[f];
+          }
+        },
+        width: 200,
+        ellipsis: {
+          showTitle: true,
+        },
+      }));
+
+      return (
+        <div css={{ overscrollBehaviorX: 'contain', overflowX: 'scroll', margin: 10 }}>
+          <span style={{ display: 'flex', justifyContent: 'center' }}>{rowPreviewMessage}</span>
+          <Table
+            columns={columns}
+            dataSource={this.state.data}
+            pagination={false}
+            sticky
+            scroll={{ x: 'min-content', y: true }}
+          />
+        </div>
+      );
+    } else {
+      return (
+        <div className='ShowArtifactPage'>
+          <div className='text-area-border-box'>{this.state.text}</div>
+        </div>
+      );
+    }
+  }
+
+  /** Fetches artifacts and updates component state with the result */
+  fetchArtifacts() {
+    const artifactLocation = getSrc(this.props.path, this.props.runUuid);
+    this.props
+      .getArtifact(artifactLocation)
+      .then((text) => {
+        try {
+          const result = Papa.parse(text, { header: true });
```

Comment:
Is it possible to parse only the first `MAX_ROW_LENGTH` rows?

Reply:
This has been resolved by using a `preview` config option in Papaparse which streams each row and aborts the serialization of the rows once the row limit is met. You should notice a performance boost. 

Perhaps for a second iteration of this feature, we can avoid downloading the whole CSV and stream the file’s bytes and parse the rows as we go.